### PR TITLE
Implement string insert command and relative-qualified ensemble dispatch

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

- Implement `string insert` command to insert a substring at a specified character index, with proper handling of edge cases (negative indices, out-of-bounds, empty strings)
- Add support for relative-qualified ensemble dispatch (e.g., `tcl::string::insert`) to resolve commands from the global namespace as if they were fully-qualified (e.g., `::tcl::string::insert`)
- Add comprehensive test coverage for `string insert` with various index forms (integer, `end`, `end-N`) and edge cases
- Update test baselines reflecting improved string test pass rate (378→404 passing tests)

### Runtime Implementation
- **`tcl_string.zig`**: Added `string_insert` function that:
  - Inserts a substring at a clamped character index (range `[0..len]`)
  - Handles `end` keyword to append after the last character (unlike `string index` where `end` is the last character)
  - Supports `end-N` syntax for relative positioning
  - Returns the insert string when target is empty, and the original string when insert is empty
  
- **`tcl_interp.zig`**: Added `try_ensemble_rewrite_relative` function to:
  - Detect relative-qualified command names (e.g., `tcl::string::insert`)
  - Rewrite them to fully-qualified form for ensemble dispatch
  - Enables upstream test suites that invoke `tcl::string::*` directly to work correctly

### Command Registration
- **`string.py`**: Registered `string insert` subcommand with WASM runtime import configuration
- **`tcl_runtime.zig`** and **`string.zig`**: Exported and wired the `string_insert` function into the command dispatch

### Test Coverage
- Added 10 new integration tests covering:
  - Insertion at start, middle, and end positions
  - `end` and `end-N` index forms
  - Negative and out-of-bounds index clamping
  - Empty target and insert string handling
  - Relative-qualified command dispatch (`tcl::string::insert`, `tcl::string::reverse`)
- Added 3 unit tests in `test_tcl_string_ops.zig` for core functionality

## Validation

- All new tests pass (10 integration tests + 3 unit tests)
- String test suite baseline improved: 378→404 passing tests (26 additional tests now pass)
- Existing string command tests continue to pass
- No regressions in other command categories

https://claude.ai/code/session_0142dMcae3KxToaBSMzXkYEj

### CODE REVIEW

File: runtime/zig/cmds/string.zig [hunk 1]
```diff
@@ -352,6 +352,7 @@ pub fn eval(words: []const i32) result_mod.InterpResult {
     if (str_eq(sp, sub.len, "reverse")) return result_mod.from_globals(rt.string_reverse(words[2]));
     if (str_eq(sp, sub.len, "repeat") and words.len >= 4) return result_mod.from_globals(rt.string_repeat(words[2], words[3]));
     if (str_eq(sp, sub.len, "replace") and words.len >= 6) return result_mod.from_globals(rt.string_replace(words[2], words[3], words[4], words[5]));
+    if (str_eq(sp, sub.len, "insert") and words.len >= 5) return result_mod.from_globals(rt.string_insert(words[2], words[3], words[4]));
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply runtime/zig/cmds/string.zig:361]: Fixed in `f9301fd` — added an `insert` entry to `sub_arity_table` (min_words=5, max_words=5), so `string insert s 1 x extra` now raises `wrong # args: should be "string insert string index insertString"` via the existing arity guard before reaching the dispatch branch.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/zig/interp/tcl_interp.zig [hunk 2]
```diff
@@ -746,6 +757,48 @@ fn try_ensemble_rewrite(cmd_s: anytype, words: []const i32) ?i32 {
...
+        }
+    }
+    if (ens_end == 0 or ens_end + 2 >= cmd_s.len) return null;
+    const ens_ptr = cmd_s.ptr + prefix.len;
+    const ens_len: u32 = ens_end - prefix.len;
+    const sub_ptr = cmd_s.ptr + ens_end + 2;
+    const sub_len: u32 = cmd_s.len - ens_end - 2;
+
+    const handler = cmd_table.lookup(ens_ptr, ens_len) orelse return null;
+
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply runtime/zig/interp/tcl_interp.zig:800]: Considered, deferred. `try_ensemble_rewrite_relative` mirrors the existing `try_ensemble_rewrite` behavior for the `::tcl::ENSEMBLE::SUBCMD` form (line 707, pre-existing) — neither validates the SUBCMD before re-dispatching. In both cases the ensemble handler raises `bad subcommand` for unknown subcommands; the failure mode for `tcl::string::nosuch` after this PR is identical to `::tcl::string::nosuch` before this PR.

Re. "currently can even fall through as success": that's a real pre-existing issue in the `string` ensemble handler, not specific to the relative form — `eval()` in `cmds/string.zig` returns `from_globals(0)` (empty result) when no subcommand branch matches. That's worth fixing, but as a separate task that touches the ensemble dispatch surface uniformly, not just the relative form.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: runtime/zig/interp/tcl_interp.zig [hunk 3]
```diff
@@ -746,6 +757,48 @@ fn try_ensemble_rewrite(cmd_s: anytype, words: []const i32) ?i32 {
...
+    const total: u32 = @as(u32, @intCast(words.len)) + 1;
+    const buf = obj_mod.alloc(total * 4);
+    const slot: [*]i32 = @ptrFromInt(buf);
+    slot[0] = obj_mod.obj_new_string(@bitCast(ens_ptr), @bitCast(ens_len));
+    slot[1] = obj_mod.obj_new_string(@bitCast(sub_ptr), @bitCast(sub_len));
+    var k: u32 = 1;
+    while (k < words.len) : (k += 1) {
+        slot[k + 1] = words[k];
+    }
+    return handler(slot[0..total]).value;
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply runtime/zig/interp/tcl_interp.zig:810]: This matches the existing `try_ensemble_rewrite` (line 707 above) byte-for-byte — same `obj_mod.alloc(total * 4)` slot array + same two synthetic `obj_new_string` headers, neither released. The runtime uses a bump allocator for these short-lived dispatch scratch slots; the comment on the original (line 735–736: *"Use the bump allocator — the slice lives only for the handler call"*) documents the convention.

If this is a real leak (i.e. the bump allocator doesn't reset between dispatches), it's a pre-existing issue that affects both helpers equally, not a new one introduced here. Switching just the relative-form helper to a different ownership model would diverge from the canonical pattern. Better tackled as a single follow-up that fixes both, with profiling to confirm whether the bump allocator's scope-end actually reclaims these.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/baselines/tcl9-tcltest-wasm/categories/mathop.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # mathop.test — auto-generated WASM baseline.
-# bucket = 'W0-runtime-error'
-# reason = 'tcl trap: site=69 unknown namespace in import pattern "::tcl::mathop::*"'
+# bucket = 'W0-timeout'
+# reason = 'exceeded wall-clock — likely infinite loop or hang'
 trap_allowed = true
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply tests/baselines/tcl9-tcltest-wasm/categories/mathop.toml:4]: Investigated — bisected the hang to **`mathop-24.5` (large shift)**. The bundle reaches `mathop-24.4` cleanly (170 test IDs failing at the watchdog cutoff with first=`mathop-1.1`, last=`mathop-24.3`). The very next test runs `[<< 1 2147483648]`, `[<< $small $big]` (where `$big = 12135435435354435435342423948763867876`), and similar large-shift forms that should error with `integer value too large to represent` per upstream — our runtime accepts the shift and goes to allocate / loop. Increasing the watchdog to 600 s confirmed it's a true hang, not slow execution.

Not a regression introduced here. Before this PR, `mathop` trapped at `namespace import ::tcl::mathop::*` (an unrelated, also-real bug fixed by #367's lazy-namespace-materialisation). With #367 + the relative-ensemble rewrite landing here, the bundle now actually executes, and a separate pre-existing arithmetic-overflow bug becomes reachable. The bucket flip surfaces an issue that was previously masked.

Per the README's "very high" leverage rating, this is a Tier-3 row that warrants its own focused PR (the fix sits in `runtime/zig/cmds/tcl_arith.zig` shift handlers — bound the count to a sane max, raise `integer value too large to represent` instead of attempting the shift). I'm not going to land that here because it touches an orthogonal subsystem and would balloon the diff. Filing as a follow-up; happy to keep this PR open until that lands if you'd rather not absorb the bucket flip until then.

---
_Generated by [Claude Code](https://claude.ai/code)_
